### PR TITLE
drivers: flash: unify andestech,qspi-nor compatible string

### PR DESCRIPTION
**Pull request by jimmyzhe**
Following the discussion in https://github.com/zephyrproject-rtos/zephyr/pull/108685#discussion_r3240196801.

Since flash_andes_qspi.c and flash_andes_qspi_xip.c both target the same hardware (Andestech ATCSPI200 controller + Flash), and XIP is a build system option and not suitable for a devicetree compatible string (`andestech,qspi-nor-xip`).

This PR replaces the `andestech,qspi-nor-xip` compatible string with `andestech,qspi-nor`.
The selection between CONFIG_FLASH_ANDES_QSPI and CONFIG_FLASH_ANDES_QSPI_XIP is now handled via Kconfig logic by checking the zephyr,flash chosen node and the XIP configuration.

This is based on the property `spi-frequency` added in the ongoing PR #108685. I will mark this PR as ready after #108685 is concluded.

### CODE REVIEW

File: drivers/flash/Kconfig.andes [hunk 1]
```diff
@@ -5,9 +5,13 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
+DT_CHOSEN_ZEPHYR_FLASH := zephyr,flash
+DT_CHOSEN_ZEPHYR_FLASH_PATH := $(dt_chosen_path,$(DT_CHOSEN_ZEPHYR_FLASH))
+DT_COMPAT_ANDESTECH_QSPI_NOR := andestech,qspi-nor
```
Comment (JarmouniA) [inline drivers/flash/Kconfig.andes]:
```suggestion
```
provided by build system

File: drivers/flash/Kconfig.andes [hunk 2]
```diff
@@ -5,9 +5,13 @@
...
+
 menuconfig FLASH_ANDES_QSPI
 	bool "Andes FLASH driver"
-	default y
+	default y if ! (XIP && $(dt_node_has_compat,$(DT_CHOSEN_ZEPHYR_FLASH_PATH),$(DT_COMPAT_ANDESTECH_QSPI_NOR)))
```
Comment (niedzwiecki-dawid) [inline drivers/flash/Kconfig.andes]: Shouldn't `!` be next to the `XIP`?

**Review by JarmouniA — COMMENTED** [review]
The driver has another problem as explained in https://github.com/zephyrproject-rtos/zephyr/pull/108685#discussion_r3234138795 and https://github.com/zephyrproject-rtos/zephyr/pull/108685#discussion_r3239907222
Also, it's missing testing in-tree (build test at the minimum).

File: drivers/flash/Kconfig.andes (hunk 2 above)
Comment (JarmouniA) [inline drivers/flash/Kconfig.andes]:
```suggestion
	default y
	depends on !FLASH_ANDES_QSPI_XIP 
```

File: drivers/flash/Kconfig.andes [hunk 3]
```diff
@@ -43,8 +47,8 @@ endif
 
 config FLASH_ANDES_QSPI_XIP
 	bool "Andes FLASH driver for XIP chip"
-	default y
-	depends on DT_HAS_ANDESTECH_QSPI_NOR_XIP_ENABLED
+	default y if $(dt_node_has_compat,$(DT_CHOSEN_ZEPHYR_FLASH_PATH),$(DT_COMPAT_ANDESTECH_QSPI_NOR))
```
Comment (JarmouniA) [inline drivers/flash/Kconfig.andes]:
```suggestion
	default y
	depends !FLASH_ANDES_QSPI
	depends on $(dt_node_has_compat,$(DT_CHOSEN_ZEPHYR_FLASH_PATH),$(DT_COMPAT_ANDESTECH_QSPI_NOR))
```

File: drivers/flash/Kconfig.andes (hunk 2 above)
Comment (jimmyzhe) [reply drivers/flash/Kconfig.andes]: I think the constraint for flash_andes_qspi.c (CONFIG_FLASH_ANDES_QSPI) is that only if **CONFIG_XIP=y** and the **`zephyr,flash` chosen node is compatible with `andestech,qspi-nor`**, we need to switch to use flash_andes_qspi_xip.c (CONFIG_FLASH_ANDES_QSPI_XIP) in this case.

If CONFIG_XIP=y but `zephyr,flash` points to a different ROM device (not compatible with `andestech,qspi-nor`), the flash_andes_qspi.c driver can still be used for other non-XIP used flash node.
For example:
```
/ {
	chosen {
		zephyr,flash = &flash0;
		...
	};
};

&spi0 {
	/* Used for XIP execution */
	flash0: flash@0 {
		compatible = "jedec,spi-nor";
		...
	};
};

&spi1 {
	/* Used as a flash driver, coexisting with CONFIG_XIP=y */
	mx25u16: qspi-nor-flash@0 {
		compatible = "andestech,qspi-nor";
		...
	};
};
```
So, I keep the current implementation to make the default values for CONFIG_FLASH_ANDES_QSPI/CONFIG_FLASH_ANDES_QSPI_XIP mutually exclusive. Based on **CONFIG_XIP=y** and the **`zephyr,flash` chosen node is compatible with `andestech,qspi-nor`**, like:
```
menuconfig FLASH_ANDES_QSPI
	bool "Andes FLASH driver"
	default y if !(XIP && $(dt_node_has_compat,$(DT_CHOSEN_ZEPHYR_FLASH_PATH),$(DT_COMPAT_ANDESTECH_QSPI_NOR)))
	...
config FLASH_ANDES_QSPI_XIP
	bool "Andes FLASH driver for XIP chip"
	default y if $(dt_node_has_compat,$(DT_CHOSEN_ZEPHYR_FLASH_PATH),$(DT_COMPAT_ANDESTECH_QSPI_NOR))
	...
	depends on XIP
```
(Note: FLASH_ANDES_QSPI_XIP is already depends on XIP, so no need to add XIP for default value expression.)

File: drivers/flash/Kconfig.andes (hunk 1 above)
Comment (jimmyzhe) [reply drivers/flash/Kconfig.andes]: Fixed it. Thanks!

**Comment by jimmyzhe** [comment]
> The driver has another problem as explained in [#108685 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/108685#discussion_r3234138795) and [#108685 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/108685#discussion_r3239907222) Also, it's missing testing in-tree (build test at the minimum).

Thanks for point this out.
Currently, we lack CI build for both flash_andes_qspi.c and flash_andes_qspi_xip.c.
I added these two test cases for CI build:
- drivers.flash.common.andestech_qspi_nor
- drivers.flash.common.andestech_qspi_nor_xip

If I missed anything from our previous discussion, please let me know. Thanks!

File: drivers/flash/Kconfig.andes (hunk 2 above)
Comment (niedzwiecki-dawid) [reply drivers/flash/Kconfig.andes]: I see your point, thanks for the explanation.

File: drivers/flash/Kconfig.andes [hunk 4]
```diff
@@ -43,8 +46,8 @@ endif
...
```
Comment (niedzwiecki-dawid) [inline drivers/flash/Kconfig.andes]: I think `$(dt_node_has_compat,$(DT_CHOSEN_ZEPHYR_FLASH_PATH),$(DT_COMPAT_ANDESTECH_QSPI_NOR))` should go to depend, not default. It doesn't make sense if the Zephyr flash chosen not is not Andes QSPI NOR

**Review by niedzwiecki-dawid — COMMENTED** [review]
While IMO this is improvement, the problem I see with such approach is that the XIP and not XIP drivers can not be used together. I'll try to improve this once this PR is merged.

File: drivers/flash/Kconfig.andes [hunk 5]
```diff
@@ -5,9 +5,12 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
+DT_CHOSEN_ZEPHYR_FLASH := zephyr,flash
+DT_CHOSEN_ZEPHYR_FLASH_PATH := $(dt_chosen_path,$(DT_CHOSEN_ZEPHYR_FLASH))
+
 menuconfig FLASH_ANDES_QSPI
 	bool "Andes FLASH driver"
-	default y
+	default y if !(XIP && $(dt_node_has_compat,$(DT_CHOSEN_ZEPHYR_FLASH_PATH),$(DT_COMPAT_ANDESTECH_QSPI_NOR)))
```
Comment (JarmouniA) [inline drivers/flash/Kconfig.andes]:
```suggestion
	default y
```

**Review by niedzwiecki-dawid — APPROVED** [review]

**Comment by jimmyzhe** [comment]
Update Kconfig logic to make CONFIG_FLASH_ANDES_QSPI_XIP strictly depend on:
- CONFIG_XIP=y
- The `zephyr,flash` chosen node being compatible with `andestech,qspi-nor`
- CONFIG_FLASH_ANDES_QSPI=**n**

This change ensure that CONFIG_FLASH_ANDES_QSPI and CONFIG_FLASH_ANDES_QSPI_XIP does not coexist.

And application is now responsible for enabling the appropriate driver (CONFIG_FLASH_ANDES_QSPI or CONFIG_FLASH_ANDES_QSPI_XIP) based on its requirements.

File: drivers/flash/Kconfig.andes (hunk 4 above)
Comment (jimmyzhe) [reply drivers/flash/Kconfig.andes]: Fixed it. Thanks!

File: drivers/flash/Kconfig.andes (hunk 2 above)
Comment (jimmyzhe) [reply drivers/flash/Kconfig.andes]: To avoid dependency loop, I only make CONFIG_FLASH_ANDES_QSPI_XIP depend on (!FLASH_ANDES_QSPI).

File: drivers/flash/Kconfig.andes (hunk 3 above)
Comment (jimmyzhe) [reply drivers/flash/Kconfig.andes]: Fixed it. Thanks!

File: drivers/flash/Kconfig.andes (hunk 5 above)
Comment (jimmyzhe) [reply drivers/flash/Kconfig.andes]: Fixed it.

**Review by JarmouniA — APPROVED** [review]
